src/db/row.rs

```diff
@@ -1,58 +1,136 @@
+                Ok(State::Exists(record_id, RecordRow::new(tx, row_id)))
             }
+            None => match record_id.resolve() {
+                Ok(Either::Left(alias)) => {
+                    tx.commit()?;
+                    Ok(State::UndefinedAlias(alias))
+                }
+                Ok(Either::Right(remote_id)) => {
+                    debug!("Checking null row for '{remote_id}'");
+                    match get_null_row_id(&tx, &remote_id)? {
+                        Some(row_id) => {
+                            Ok(State::NullRemote(remote_id, NullRecordRow::new(tx, row_id)))
+                        }
+                        None => Ok(State::UndefinedRemote(remote_id, Undefined::new(tx))),
+                    }
+                }
+                Err(record_error) => {
+                    tx.commit()?;
+                    Ok(State::InvalidRemoteId(record_error))
+                }
+            },
         }
     }
 }
 
-/// A private trait which allows the caller to prepare (cached) sql statements.
-trait DatabaseTransaction {
-    fn prepare(&self, sql: &str) -> Result<rusqlite::Statement, rusqlite::Error>;
+/// A representation of the database state beginning with an arbitrary [`RemoteId`].
+#[derive(Debug)]
+pub enum RemoteState<'conn> {
+    /// The `Records` row exists.
+    Exists(RemoteId, RecordRow<'conn>),
+    /// The `Records` row does not exist and the `NullRecords` row exists.
+    NullRemote(RemoteId, NullRecordRow<'conn>),
+    /// The `Records` and `NullRecords` rows do not exist.
+    UndefinedRemote(RemoteId, Undefined<'conn>),
+    /// The remote id is invalid.
+    InvalidRemoteId(RecordError),
+}
 
-    fn prepare_cached(&self, sql: &str) -> Result<rusqlite::CachedStatement, rusqlite::Error>;
+impl<'conn> RemoteState<'conn> {
+    /// Determine the current state of the database, as corresponds to the provided record
+    /// identifier.
+    #[inline]
+    pub fn determine(tx: Transaction<'conn>, remote_id: RemoteId) -> Result<Self, rusqlite::Error> {
+        match get_row_id(&tx, &remote_id)? {
+            Some(row_id) => {
+                debug!("Beginning new transaction for row '{row_id}'.");
+                Ok(Self::Exists(remote_id, RecordRow::new(tx, row_id)))
+            }
+            None => {
+                debug!("Checking null row for '{remote_id}'");
+                match get_null_row_id(&tx, &remote_id)? {
+                    Some(row_id) => Ok(Self::NullRemote(remote_id, NullRecordRow::new(tx, row_id))),
+                    None => Ok(Self::UndefinedRemote(remote_id, Undefined::new(tx))),
+                }
+            }
+        }
+    }
 }
 
-/// A representation of a row in the `Records` table which is missing.
+/// The database state when:
+/// 1. There is no row in `Records`.
+/// 2. There is no row in `NullRecords`.
 #[derive(Debug)]
-pub struct MissingRecordRow<'conn> {
+pub struct Undefined<'conn> {
```

Comment:
Better name for `Undefined`?

Reply:
This is a hard one. `Idle`?

Reply:
Actually, maybe `MissingRow` is quite precise here?